hkmc2/shared/src/test/mlscript-compile/Rendering.mls

```diff
@@ -35,27 +35,350 @@ fun interleave(sep)(...args) =
     if i < len do set res.[idx + 1] = sep
   res
 
-fun render(arg) = if
-  arg is
-    undefined then "undefined"
-    null      then "null"
-    Array     then fold(+)("[", ...interleave(", ")(...map(render)(...arg)), "]")
-    Str       then JSON.stringify(arg)
-    Set       then fold(+)("Set{", ...interleave(", ")(...map(render)(...arg)), "}")
-    Map       then fold(+)("Map{", ...interleave(", ")(...map(render)(...arg)), "}")
-    Function  and
-      let p = Object.getOwnPropertyDescriptor(arg, "prototype")
-      (p is Object and p.("writable")) || (p is undefined) then
-        "[function" + (if arg.name is
-          ""  then ""
-          nme then " " + nme
-        ) + "]"
-    Object then
-      if arg.constructor.name is "Object"
-      then
-        let es = Object.entries(arg)
-        fold(+)("{", ...interleave(", ")(...map(case [k, v] then k + ": " + render(v))(...es)), "}")
-      else String(arg)
-  let ts = arg.("toString") // not accessing as `arg.toString` to avoid the sanity check
-  ts is undefined then "[" + typeof(arg) + "]"
-  else ts.call(arg)
+// * Apply `f` to `xs` and build a string from non-null results using the given
+// * separator and starting/ending symbols. If there is no non-null results,
+// * return `empty` instead.
+fun mkString(xs, f, start, sep, end, empty) =
+  let
+    it = xs.(Symbol.iterator)()
+    res = it.next()
+    buffer = null
+  // Consume the iterator until we find the first non-null value.
+  while res.done is false and buffer is null do
+    if f(res.value) is
+      Str as str do set buffer = start + str
+    set res = it.next()
+  // If the buffer is still null, this means we have drained the iterator.
+  if buffer is null then empty else...
+  while res.done is false do
+    let value = f(res.value)
+    if value is Str as s do
+      set { buffer += sep, buffer += s }
+    set res = it.next()
+  set buffer += end
+  buffer
+
+// Do not wrap property names in quotes if they match the following pattern.
+let identifierPattern = new RegExp of
+  """^[\p{ID_Start}_$][\p{ID_Continue}_$]*$""", "u"
+
+fun noNeedForQuotes(id) = identifierPattern.test(id)
+
+let MIN_BREAK_LENGTH = 60
+
+let NULL_PROTO = "[Object: null prototype] "
```

Comment:
Why use this inconsistent printing syntax? We don't want to copy JS, here.

Can we make it
```suggestion
let NULL_PROTO = ""
```

and print objects with the `Object` prototype as `Object { ... }`?

Reply:
Good idea!

Reply:
Ok, so you reverted that idea. What's the current plan for constructor-less objects?

Reply:
Yes, I did. You told me to make the rendering output more MLscripty. I think the idea of “objects whose prototype is `null`” only exists in JavaScript. So I no longer plan to distinguish between objects created using object literals (which should be the same as what you call constructor-less objects) and objects with a `null` prototype.

Reply:
(If we do this, then we'll want to display `Object` instances as before your latest change.)

Reply:
What, that's crazy! Thanks for benchmarking it.

I'm guessing the null-prototype objects are not treated as class instances by the JIT (which is how JITs represent objects efficiently).